posts/2023-06-12-streams-in-scala.adoc

```diff
@@ -0,0 +1,214 @@
+As long as there is outstanding demand, the publisher can continue pushing data to the subscriber as it becomes available. However, when the demand is exhausted, the publisher cannot send data unless prompted by a demand signal from downstream. This mechanism, known as backpressure, ensures controlled flow and prevents overwhelming the subscriber. In response to backpressure, the source can choose to allocate more resources, slow down its production, or even discard data.
+
+image::../media/2023-06-12-streams-in-scala/reactive-stream-representation.png[]
+
+To summarise, handling an un-bounded volume of data in an asynchronous system requires some form of control between the producer and the consumer otherwise we would have overwhelming data sent to the consumer from multiple threads. The Reactive stream standard Introduces a concept of back-pressure which is a means of communication between the producers and the consumer. The reactive stream defines an interface which must be implemented.
+
+The low-level interface of the Reactive streams:
+
+[source,scala]
+----
+trait Publisher[T] {
+ def subscribe(s: Subscriber[T]): Unit
+}
+
+trait Subscriber[T] {
+  def onSubscribe(s: Subscription): Unit
+  def onNext(t: T): Unit
+  def onError(t: Throwable): Unit
+  def onComplete(): Unit
+}
+
+trait Subscription {
+  def request(n: Int): Unit
+  def cancel(): Unit
+}
+----
+
+The below libraries take into account this reactive stream and implement high-level stream API
+
+=== https://doc.akka.io/docs/akka/current/stream/index.html[Akka Streams]:
+Akka Streams is a powerful and scalable stream processing library built on top of the Akka toolkit. It provides a high-level DSL for composing and executing stream-based computations. Akka Streams offers backpressure support, fault-tolerance, and integration with other Akka components. It's widely used in building reactive and distributed systems.
+
+=== https://fs2.io/#/getstarted/example[fs2]:
+fs2 (Functional Streams for Scala) formally called Scalaz-Stream is a functional stream processing library that provides a purely functional, composable, and resource-safe approach to handling streams. It leverages functional programming concepts such as cats-effect and functional abstractions to build complex stream processing pipelines. fs2 focuses on efficiency, type safety, and composability. Beyound stream processing, fs2 can be used for https://underscore.io/blog/posts/2018/03/20/fs2.html[everything] from task execution to control flow.
+
+=== https://monix.io/docs/current/intro/hello-world.html[Monix]:
+Monix is a library that provides reactive programming and asynchronous streams for Scala. It offers various abstractions like Observable, Iterant, and Task to handle asynchronous computations, data streams, and reactive programming. Monix supports backpressure and provides powerful operators for manipulating and transforming streams.
```

Comment:
While Monix has some features that are stream-like, I'm not actually sure I'd describe it as a streaming library. It's for sure event based both focuses on other things as well. I'm not sure, take it or leave it.